python/mlc_chat/model/llama/llama_model.py

```diff
@@ -104,6 +107,30 @@ def forward(self, x: Tensor):
         return self.down_proj(op.silu(x1) * x2)
 
 
+@dataclasses.dataclass
```

Comment:
I'd love to argue against the idea of having nullable-simulated tagged unions in the code base to try to merge code path, neither in `AttentionArgs` nor in switching in-between `input_ids` and `input_embeds`. I understand that getting rid of them may lead to a few lines of code duplication in interface logics, but on the other hand, it will improve the readability significantly, where different code will not cross over.

Reply:
I initially proposed `AttentionArgs` as eventually we would like to move away from the single batch impl possible and into the common kv interface.

`input_ids` and `input_embeds` should be unified into always use embedding later.

The question is what do we do in the meantime, while i think either way is ok. perhaps we can go with two code path for now..

Reply:
Updated `llama_model.py` with explicit `forward_single_seq` and `forward_batch_seq`. Right now codepaths are barely shared (and consequently, have certain amount of duplication). Would you mind taking another look?

Reply:
one minor nit: how about we do batch_xxx (batch as prefix ) so this way methods can be grouped easily. Otherwise lgtm

Reply:
Looks pretty clear to me now!

Reply:
Updated just now :-)